script/build.sh

```diff
@@ -3,44 +3,100 @@
 set -e
 
 DIR="$(cd "$(dirname "$0")/../" && pwd)"
+INCLUDE_DIR=$DIR
+SOURCE_DIR=$DIR
+BUILD_DIR=$DIR/build
+BUILD_INT_DIR=$BUILD_DIR/intermediate
+BUILD_BIN_DIR=$BUILD_DIR/bin
+STATIC_BUILD_INT_DIR=$BUILD_DIR/intermediate/static
+SHARED_BUILD_INT_DIR=$BUILD_DIR/intermediate/shared
+BUILD_LIB_DIR=$BUILD_DIR/lib
+STATIC_BUILD_LIB_DIR=$BUILD_LIB_DIR/static
+SHARED_BUILD_LIB_DIR=$BUILD_LIB_DIR/shared
+TEST_SOURCE_DIR=$DIR/test
+TEST_INT_DIR=$BUILD_INT_DIR/test
+TEST_BIN_DIR=$BUILD_BIN_DIR/test
+EXAMPLES_SOURCE_DIR=$DIR/examples
+EXAMPLES_INT_DIR=$BUILD_INT_DIR/examples
+EXAMPLES_BIN_DIR=$BUILD_BIN_DIR/examples
+
+CFLAGS="-I$INCLUDE_DIR"
+CXXFLAGS="-I$INCLUDE_DIR"
+LDFLAGS=""
 
 if [ "$(uname)" = "Darwin" ]; then
-	FLAGS="-DWEBVIEW_COCOA -std=c++11 -Wall -Wextra -pedantic -framework WebKit"
+	LDFLAGS="-framework WebKit"
+	CXXFLAGS="$CXXFLAGS -DWEBVIEW_COCOA -std=c++11 -Wall -Wextra -pedantic"
 else
-	FLAGS="-DWEBVIEW_GTK -std=c++11 -Wall -Wextra -pedantic $(pkg-config --cflags --libs gtk+-3.0 webkit2gtk-4.0)"
+	LDFLAGS="$(pkg-config --libs gtk+-3.0 webkit2gtk-4.0)"
+	CXXFLAGS="$CXXFLAGS -DWEBVIEW_GTK -std=c++11 -Wall -Wextra -pedantic $(pkg-config --cflags gtk+-3.0 webkit2gtk-4.0)"
 fi
 
 if command -v clang-format >/dev/null 2>&1 ; then
 	echo "Formatting..."
-	clang-format -i \
-		"$DIR/webview.h" \
-		"$DIR/webview_test.cc" \
-		"$DIR/main.cc"
+	# TODO: Clean up examples and remove them from the exclusion list
+	find . -type f -not -name "$EXAMPLES_SOURCE_DIR" -regextype posix-extended -iregex '.*\.(h|cc?)$' -exec clang-format -i {} +
 else
 	echo "SKIP: Formatting (clang-format not installed)"
 fi
 
 if command -v clang-tidy >/dev/null 2>&1 ; then
 	echo "Linting..."
-	clang-tidy "$DIR/main.cc" -- $FLAGS
-	clang-tidy "$DIR/webview_test.cc" -- $FLAGS
+	# TODO: Clean up examples and remove them from the exclusion list
+	find . -type f -not -name "$EXAMPLES_SOURCE_DIR" -regextype posix-extended -iregex '.*\.cc$' -exec clang-tidy {} -- $CXXFLAGS \;
 else
 	echo "SKIP: Linting (clang-tidy not installed)"
 fi
 
-echo "Building C++ example"
-c++ main.cc $FLAGS -o webview
+mkdir -p "$TEST_INT_DIR" "$TEST_BIN_DIR" \
+	"$STATIC_BUILD_INT_DIR" "$STATIC_BUILD_LIB_DIR" \
+	"$SHARED_BUILD_INT_DIR" "$SHARED_BUILD_LIB_DIR" \
+	"$EXAMPLES_INT_DIR" "$EXAMPLES_BIN_DIR"
+
+echo "Building static library"
+c++ -c "$SOURCE_DIR/webview.cc" -DWEBVIEW_BUILDING -DWEBVIEW_STATIC $CXXFLAGS -o "$STATIC_BUILD_INT_DIR/webview.o"
+ar rcs "$STATIC_BUILD_LIB_DIR/libwebview.a" "$STATIC_BUILD_INT_DIR/webview.o"
+
+echo "Building shared library"
+c++ -c "$SOURCE_DIR/webview.cc" -DWEBVIEW_BUILDING -DWEBVIEW_SHARED -fPIC -fvisibility=hidden -fvisibility-inlines-hidden $CXXFLAGS -o "$SHARED_BUILD_INT_DIR/webview.o"
```

Comment:
Exactly which flags to pass in for the shared library could be open for discussion.